### ISSUE TITLE
[`needless_late_init`] suggestion doesn't compile when the initializing takes place inside macro

### ISSUE THREAD

**Issue by J-ZhengLi** [rust-lang/rust-clippy#13776]
### Summary

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=19a2d9aae1fe6edb9f6512357cac52b5)

### Lint Name

needless_late_init

### Reproducer

I tried this code:

```rust
macro_rules! need_late_init {
    ($var:expr, $val:literal) => {
        $var = $val;
    }
}

fn main() {
    let x;
    need_late_init!(x, 10);
    assert_eq!(x, 10);
}
```

I saw this happen:

```
error: unneeded late initialization
  --> src/main.rs:6:9
   |
6  |         $var = $val;
   |         ^^^^ initialised here
...
12 |     let x;
   |     ^^^^^^ created here
13 |     
14 |     need_late_init!(x, 10);
   |     ---------------------- in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_late_init
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![deny(clippy::needless_late_init)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the macro `need_late_init` (in Nightly builds, run with -Z macro-backtrace for more info)
help: move the declaration `x` here
   |
12 ~     
13 |     
14 ~     need_late_init!(let x, 10);
   |
```

I expected to see this happen:

No warning, since `let x` is not an expr

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-pc-windows-gnu
release: 1.83.0
LLVM version: 19.1.1
```
